### DISCUSSION TITLE
Fetching a whole database

### DISCUSSION THREAD

**Discussion by louisepght** [tmijail/Odoo-Power-BI-Connector#10] [Show and tell]
Hello, 

I'm using your connector to connect Odoo to powerBI. 
I have a lot of models on Odoo (more than 200) meaning I cant use search_read for each of them. I also want to keep the links between the tables. 
In a nutshell, I want to fetch all the data and database from Odoo.
Is that possible?

**Comment by tmijail** [comment]
You'd have to write a `search_read('model.name')` for each one.

Maybe it would be possible to automate it by reading the `ir.model` table like it's done [here](https://github.com/tmijail/Odoo-Power-BI-Connector/blob/21c9779de47c719317068f65ad236717e074a624/Odoo/Odoo.pq#L421) and using `Table.TransformRows` to do a `search_read` on each model. Still, you would have to somehow convert each row in the resulting table into a separate query.

Regarding maintaining the links between tables, Odoo's and PowerBI's relationships aren't 100% compatible (especially many2many relationships). You can get PowerBI to automatically recognize one2one or one2many relationships between two tables if the related columns have the same name. For example, the `sale.order` table has a `partner_id` column which is related with the `id` column in the `res.partner` table. If you import both tables and rename the `id` column in `res.partner` to `partner_id`, PBI will see that the two tables have columns with the same name and infer the relationship.

Either way, I'd recommend you import just the models, fields and relationships necessary for your reports. Large PowerBI models are hard to mantain and your proposed model _will_ require maintenance when you upgrade to a newer Odoo version or when your user's permissions change. Also, Odoo servers are prone to time out when you try to read a lot of data at once (especially if that data includes [computed](https://www.odoo.com/documentation/15.0/developer/reference/backend/orm.html#computed-fields), non [stored](https://www.odoo.com/documentation/15.0/developer/reference/backend/orm.html#odoo.fields.Field) fields)
